### ISSUE TITLE
ServerSentEvent support for data being a multi-line string and possible other ServerSentEvent enhancements

### ISSUE THREAD

**Issue by fizzy33** [http4s/http4s#4865]
For server sent events data can be multi-line and in that case multiple `data:` lines should be sent as outlined here https://html.spec.whatwg.org/multipage/server-sent-events.html

The current implementation only sends a single `data:` line followed by the entire data string.  So if that data string 

```
final case class ServerSentEvent(
    data: String,
    eventType: Option[String] = None,
    id: Option[EventId] = None,
    retry: Option[Long] = None
) extends Renderable {
  def render(writer: Writer): writer.type = {
    writer << "data: " << data << "\n"
    eventType.foreach(writer << "event: " << _ << "\n")
    id match {
      case None =>
      case Some(EventId.reset) => writer << "id\n"
      case Some(EventId(id)) => writer << "id: " << id << "\n"
    }
    retry.foreach(writer << "retry: " << _ << "\n")
    writer << "\n"
  }

  override def toString: String =
    s"ServerSentEvent($data,$eventType,$id,$retry)"
}
```

Here is the fix we are using in production noting it does a few more things than just fix the multi-line data issue that are worth considering i.e. comments and data is not required which allow for being a more complete sse implementation.

If these look good (or just the data fix only) I am happy to put together an MR.  Noting sending comment lines is the idiomatic ping for sse so it is critical for any robust real world implementation of SSE.

data changing from String to Option[String] is likely a breaking change but a worthwhile one to do now to get a complete ServerSentEvent implementation.  The change in http4s's internals appears trivial (only a few places) clients using it though will break so it may be best to time that change with a larger version change.  

```

  final case class ServerSentEvent(
      data: Option[String] = None,
      comment: Option[String] = None,
      eventType: Option[String] = None,
      id: Option[EventId] = None,
      retry: Option[FiniteDuration] = None
  ) extends Renderable {
    def render(writer: Writer): writer.type = {
      data.foreach ( _.linesIterator.foreach { dataLine =>
        writer << "data: " << dataLine << "\n"
      })
      comment.foreach(_.linesIterator.foreach { commentLine =>
        writer << ": " << commentLine << "\n"
      })
      eventType.foreach(writer << "event: " << _ << "\n")
      id match {
        case None =>
        case Some(EventId.reset) => writer << "id\n"
        case Some(EventId(id)) => writer << "id: " << id << "\n"
      }
      retry.foreach(writer << "retry: " << _.toMillis << "\n")
      writer << "\n"
    }

    override def toString: String =
      s"ServerSentEvent($data,$eventType,$id,$retry)"
  }
```

thoughts !?!?!

**Comment by fizzy33** [comment]
Thinking about this more now.  I will prepare a simple fix for the immediate issue.

If people want a more complete / pedantic they can just cut and paste the provided `ServerSentEvent` and it should just work.  Also open to doing a more complete PR with that as well just depends on what people want.

**Comment by fizzy33** [comment]
Here is the PR for the obvious fix to the immediate issue https://github.com/http4s/http4s/pull/4866

**Comment by rossabaker** [comment]
The grammar is:

```
event         = *( comment / field ) end-of-line
```

Are these two SSEs equal, or do we need to preserve the difference?

```
: A
data: 1
: B
data: 2
```

```
: A
: B
data: 1
data: 2
```

The processing instructions say comments can be discarded, so I think we can normalize them by putting the comments up front.  I also think we can normalize them by serializing the fields in any order, and discarding duplicates of the non-data fields.  If all that is true, your model looks good.

It's weird to have the possibility of an empty event, but I think that's okay: the grammar says any number.  I think two blank lines in a row represent "the empty SSE."

A problem with the `FiniteDuration` is that it can be negative, but so can the `Long` in the current model.  I don't think it's worth refined types to eliminate, and at least yours is a more sensible type.

Using the `linesIterator` eliminates the difference between `\r`, `\n`, and `\r\n`.  `\n` is the best one to use, but it makes a lossy conversion if those delimiters are significant to the data.  I don't have enough experience with SSE to know whether that's a concern or not.

**Comment by fizzy33** [comment]
Yeah I toss and turn on the `FiniteDuration` what I cut and pasted was what we use internally and are a bit more opinionated than I think http4s should be.  The benefit is users don't have to look at the spec to see what that retry is in (milliseconds) downside is you are a bit removed from the actual protocol.

Great point on `linesIterator`.  I think we should pass on whatever we get.  Noting having done a quick switch to `linesWithSeparators` reveals some decoder regressions, so I have a bit of work to get that plumbed through.

**Comment by rossabaker** [comment]
I think the `FiniteDuration` is better than `Long`.  I wish there were an `UnsignedFiniteDuration` (there are a few other places in HTTP it would be correct), but `FiniteDuration` is totally consistent with our other use.

**Comment by rossabaker** [comment]
0.22 and 0.23 release candidates are imminent, probably as soon as Circe is released.  If we want to get comment support, we'll need to add it to the model for bincompat reasons, or else it will have to wait for 1.0.  We might want to prioritize that over the carriage return preservation being discussed in #4866.

**Comment by rossabaker** [comment]
Mostly done in #4888, but leaving open to track the nuance of preserving carriage returns.

**Comment by fizzy33** [comment]
Now that we have added comment to ServerSentEvent we will need to apply the multi-line fix to that as well.

Noting this is mainly waiting on https://github.com/typelevel/fs2/pull/2418 to work it's way out.  As one can't properly decode a server sent events response / protocol without the actual line endings.  The fact that we assume a line is terminated by \n without really knowing causes some corner cases in the unit tests that will get cleaned up if/when we switch to the ultimate fs2.text.linesWithEndings solution
